### PR TITLE
feat: get list of `ConnectionInfo`s or an individual node's `ConnectionInfo`

### PR DESCRIPTION
**Pull request by ramfox**
## Description

Display connection information for all nodes we have connected with or have supplied our node with addresses for.

Display connection information on a particular node, using its node id.

## Change checklist

- [x] Self-review.
- [x] Documentation updates if relevant.

### CODE REVIEW

**Comment by b5** [comment]
ZOMG I WANT THIS.

Howeve, given that you had to censor the screenshot, I'd say right off the bat we should go with listing by `NodeID`, and omitting IP addresses. wdyt?

**Comment by ramfox** [comment]
> ZOMG I WANT THIS.
> 
> Howeve, given that you had to censor the screenshot, I'd say right off the bat we should go with listing by `NodeID`, and omitting IP addresses. wdyt?

Screen Shot 2023-08-30 at 22 04 06

What do you think is the "right" language when we don't have any latency info?

**Comment by b5** [comment]
Can we use short NodeIDs, like we do for doc & author ids in the console? We could also do this in a follow-up.

> What do you think is the "right" language when we don't have any latency info?

`with a latency of Unknown`, using "unknown" is the right thing to do here imho

File: iroh/src/commands.rs [hunk 1]
```diff
@@ -318,6 +321,15 @@ pub enum NodeCommands {
 impl NodeCommands {
     pub async fn run(self, client: RpcClient) -> Result<()> {
         match self {
+            Self::Connections => {
+                let mut connections = client.server_streaming(ConnectionsRequest).await?;
+                println!("Connections:");
+                while let Some(Ok(Ok(ConnectionsResponse { node_info }))) = connections.next().await
```
Comment (ramfox) [inline iroh/src/commands.rs:327]: @rklaehn can I get some RPC help?

I don't understand how I ended up with double `Result`s. Did I do something incorrectly, or is expected?

**Comment by ramfox** [comment]
@b5 
Screen Shot 2023-08-30 at 22 44 39

```
Connections:
	hpf6qa4h…
	derp region: 1
	direct connection
	5175250ns latency

	nzx6mxi6…
	derp region: 1
	relay connection
	unknown latency

	xce2h4zq…
	derp region: 1
	relay connection
	unknown latency

	xuko33wo…
	derp region: 1
	direct connection
	3489375ns latency
```

If you write out one example of how you think it should look, I'll just do it, because it doesn't look right yet but i don't know how to adjust it.

**Comment by ramfox** [comment]
```
Connections:
	xuko33wo…
	derp region: 1
	connection type: direct
	latency: 3489375ns 
```
?

**Comment by b5** [comment]
zomg you fast. the problem is you're not doing one entry per line. aim for a table layout:

```
nodeId    derp region    conn type    latency
xuko33wo…  1             direct       56.5678ms
```

Huge win here is each row will be a constant-ish length except for the last value (latency), which makes the formatting much easier, just use `\t` characters between values.

latency should be displayed in `ms` not `ns`.

**Comment by ramfox** [comment]
Screen Shot 2023-08-30 at 23 37 51

will look at other ways to display latency tomorrow!

File: iroh/src/commands.rs (hunk 1 above)
Comment (Frando) [reply iroh/src/commands.rs:326]: This is expected behavior. The first `Result` is around a connection error while reading the response, and the second `Result` is the `RpcResult` that you are sending as `Item` in the stream.

**Review by Frando — COMMENTED** [review]
Looks good to me!

Maybe let's add a `node connection NODE_ID` command as well that shows all details for a single node id, including IP addresses. This info is available at the node, so I think it makes sense to expose it in some way. Useful for debugging etc.

**Comment by dignifiedquire** [comment]
https://github.com/manlge/human-time makes nicer timestamps

**Comment by dignifiedquire** [comment]
https://github.com/Nukesor/comfy-table can render easy very nice tables :)

**Comment by dignifiedquire** [comment]
> Can extend this functionality to the console.

lets add this to the repl for sure

**Comment by dignifiedquire** [comment]
We should add this to the `client` api as well, this will be very useful from the API

**Comment by ramfox** [comment]
### `iroh node connections`
CLI
Screen Shot 2023-08-31 at 23 15 21

Console
Screen Shot 2023-08-31 at 23 15 59

Adjusted the `node id` field to show the full ID, otherwise you wouldn't be able to fetch any individual node connection information with `iroh node connection NODE_ID`

### `iroh node connection NODE_ID`
CLI
Screen Shot 2023-08-31 at 23 15 31

Console
Screen Shot 2023-08-31 at 23 16 10

**Review by dignifiedquire — APPROVED** [review]

**Pull request merged**